beacon-chain/blockchain/forkchoice_update_execution.go

```diff
@@ -40,28 +42,57 @@ func (s *Service) getStateAndBlock(ctx context.Context, r [32]byte) (state.Beaco
 	return headState, newHeadBlock, nil
 }
 
-func (s *Service) forkchoiceUpdateWithExecution(ctx context.Context, newHeadRoot [32]byte) error {
+func (s *Service) forkchoiceUpdateWithExecution(ctx context.Context, newHeadRoot [32]byte, proposingSlot primitives.Slot) error {
 	isNewHead := s.isNewHead(newHeadRoot)
-	if !isNewHead && !s.isNewProposer() {
+	if !isNewHead && !s.isNewProposer(proposingSlot) {
 		return nil
 	}
 
-	headState, headBlock, err := s.getStateAndBlock(ctx, newHeadRoot)
-	if err != nil {
-		log.WithError(err).Error("Could not get forkchoice update argument")
-		return nil
+	var headState state.BeaconState
+	var headBlock interfaces.ReadOnlySignedBeaconBlock
+	var headRoot [32]byte
+	var err error
+
+	shouldUpdate := isNewHead
+	if isNewHead && s.isNewProposer(proposingSlot) && !features.Get().DisableReorgLateBlocks {
+		if proposingSlot == s.CurrentSlot() {
+			proposerHead := s.ForkChoicer().GetProposerHead()
+			if proposerHead != newHeadRoot {
+				shouldUpdate = false
+			}
+		} else if s.ForkChoicer().ShouldOverrideFCU() {
+			shouldUpdate = false
+		}
+	}
+	if shouldUpdate {
+		headRoot = newHeadRoot
+		headState, headBlock, err = s.getStateAndBlock(ctx, newHeadRoot)
+		if err != nil {
+			log.WithError(err).Error("Could not get forkchoice update argument")
+			return nil
+		}
+	} else {
+		// We are guaranteed that the head block is the parent
+		// of the incoming block. We do not process the slot
+		// because it will be processed anyway in notifyForkchoiceUpdate
+		headState = s.headState(ctx)
+		headRoot = s.headRoot()
+		headBlock, err = s.headBlock()
+		if err != nil {
+			return errors.Wrap(err, "could not get head block")
+		}
 	}
```

Comment:
There are a lot of conditions here. Hard to read. I wonder if we can make it cleaner